### DISCUSSION TITLE
2.4.0版本偶发查询结果不稳定

### DISCUSSION THREAD

**Discussion by douxiaofeng99** [milvus-io/milvus#35782] [Q&A and General discussion]
搭建的2.4.0向量库，数据切片后，过后进行查询，发现查询同样的问题，偶发性的结果不一样，有时不是想要的答案。请问，对于这种情况，如何进行排查？ Collection是单独的，这时对于这个Collection也没有做任何操作

**Comment by yhmo** [comment]
- 当milus接收到insert请求时，数据并不是立即能搜索到，因为数据要先进入message queue作为write-ahead-log。当querynode异步从message queue里接收到这些数据，才会被搜索到。从数据进入message queue到querynode接收到，可能会有秒级的间隔。调用search接口时consistency_level参数可以控制数据的可见性。
- 当querynode接收到数据之后，并不会立即把数据落盘，而是累积在一个buffer里，作为growing segment存在，growing segment有临时索引(IVF_FLAT)。一般情况下，buffer会在数分钟之后才会落盘。在这数分钟里搜索到的结果可能和数分钟之后搜索到的结果略有不同，因为落盘之后的sealed segment会重新建索引。
- 落盘的sealed segment还会参与compaction，多个小segment合并成大的segment，合并之后的segment也会重新建索引。由于数据合并索引重建，合并之前的搜索结果和合并之后的搜索结果有可能会有一些细微差别。

除了以上几项有可能造成微小的搜索结果差别之外，理论上其他情况下搜索结果都是稳定的(数据不变，搜索的参数不变)，除非是有什么bug。如果你能用脚本证明查询结果偶发不一样，就把脚本发给我们。


PS: 以后在milvus社区里提问请尽量使用英语

**Reply by OpenSourceApp-coder** [reply]
没想到现代社会中国这么强大了，还有这么多崇洋媚外的人，用英语能显的你们高大上吗？使用中文影响你们成为国际化的团队了？这种看不起自己国家语言的公司早晚倒闭。

**Reply by yhmo** [reply]
@OpenSourceApp-coder 你过激了。你看我这么详细地用中文回复就知道我并没有任何看不起的意思。
有其他国家的用户也会碰到你类似的问题，如果都用不同的语言，意味着我每个人都得回复一次，并且不利于搜索统一的答案。
如果你实在不喜欢写成英文，那写中文也ok，或者进milvus的微信用户群(纯中文)。我得原话是“请尽量使用英语”，没有说必须使用英语。

**Reply by douxiaofeng99** [reply]
我是怕英语说不清楚，所以才使用了中文。刚开始我也怀疑是不是索引还没建起来。但从我们现场日志看，插入操作时在一天前完成的。或者几天都没有操作。因此，就打消了这个怀疑。另外，由于这是在生产环境偶发，我们也没办法复现。还有就是，生产环境也没有搭建集群模式，还是单机模式。所以，这里求助，有没有办法排查造成这个现象的方法。以便再次出现时，确定原因。
前天，已经将2.4.0升级到2.4.9，观察一下效果

**Reply by yhmo** [reply]
先确认一下search的入参是否有变化？比如目标向量有没有改变，过滤条件有没有改变。
使用的SDK是python还是java，版本号是？
索引是哪种索引？

**Comment by xiaofan-luan** [comment]
> 搭建的2.4.0向量库，数据切片后，过后进行查询，发现查询同样的问题，偶发性的结果不一样，有时不是想要的答案。请问，对于这种情况，如何进行排查？ Collection是单独的，这时对于这个Collection也没有做任何操作

偶发性的结果不一样的，可能是正常情况，这通常发生在
1. 数据flush时，索引从临时索引转换为最终的索引
2. compaction之后，segment会发生合并操作

对于向量数据库来讲，使用索引即导致数据不会100%命中。那么任何reindex，merge的操作都会可能会导致查询不一致。
所以可能需要考虑这个偶发的频率，另外调整index参数可以缓解这个问题
